### PR TITLE
Week 4 assignment

### PR DESCRIPTION
**Pull request by konarssuresh**
This PR contains code change for week 4 assignment questions
* creating promise function
* implementing inheritence
* fibonacci series using iterators

### CODE REVIEW

File: Week-4/fibonacci.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
+let fibonacciSeries = {
+  [Symbol.iterator]() {
+    let series = [];
+    let currentIndex;
```
Comment (divjot-singh) [inline Week-4/fibonacci.js:4]: could have initialised currentIndex to 0 here then there wouldnt have been any need for the condition in line 7

File: Week-4/fibonacci.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    return {
+      next() {
+        if (currentIndex || currentIndex === 0) currentIndex++;
+        else currentIndex = 0;
+        if (series.length === 0 || series.length === 1) {
```
Comment (divjot-singh) [inline Week-4/fibonacci.js:9]: series.length <=1

File: Week-4/promise.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  const p = new MyPromise((resolve, reject) => {
+    if (n % 5 !== 0) {
+      resolve("number is divisible by 5");
+    } else {
+      reject("number is not divisible by 5");
+    }
+  });
+};
+
+getNumber();
```
Comment (divjot-singh) [inline Week-4/promise.js:35]: this is not an ideal implementation. Your promise should fulfill the following use case:

const p = new MyPromise((resolve, reject) => {
    if (n % 5 !== 0) {
      resolve("number is divisible by 5");
    } else {
      reject("number is not divisible by 5");
    }
  });
  
  p.then((val) => { console.log(val)}).catch((error)=> {console.log(error)}) 
  
  Also, please make sure if resolve or reject is called more than once, the firs execution is taken into account.

File: Week-4/promise.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  if (this.state === "pending") {
+    this.value = value;
+    this.state = "rejected";
+  }
+};
+
+function MyPromise(moderator) {
+  this.state = "pending";
+  this.value = undefined;
+  this.resolve = resolveFunc.bind(this);
```
Comment (divjot-singh) [inline Week-4/promise.js:18]: resolve and reject should be private variables

**Review by divjot-singh — CHANGES_REQUESTED** [review]

File: Week-4/fibonacci.js (hunk 1 above)
Comment (konarssuresh) [reply Week-4/fibonacci.js]: Fixed.initialized currentIndex to 0

File: Week-4/fibonacci.js (hunk 2 above)
Comment (konarssuresh) [reply Week-4/fibonacci.js]: fixed

File: Week-4/promise.js (hunk 3 above)
Comment (konarssuresh) [reply Week-4/promise.js:116]: implemented promise function as we discussed. Let me know if my current implementation is not proper

**Review by divjot-singh — APPROVED** [review]

**Pull request merged**